cpp/bench/ann/src/faiss/faiss_gpu_wrapper.h

```diff
@@ -225,19 +251,79 @@ void FaissGpu<T>::search(const T* queries,
 
-  if (this->refine_ratio_ > 1.0) {
-    // TODO: FAISS changed their search APIs to accept the search parameters as a struct object
-    // but their refine API doesn't allow the struct to be passed in. Once this is fixed, we
-    // need to re-enable refinement below
-    // index_refine_->search(batch_size, queries, k, distances,
-    // reinterpret_cast<faiss::idx_t*>(neighbors), this->search_params_.get()); Related FAISS issue:
-    // https://github.com/facebookresearch/faiss/issues/3118
-    throw std::runtime_error(
-      "FAISS doesn't support refinement in their new APIs so this feature is disabled in the "
-      "benchmarks for the time being.");
+  if (refine_ratio_ > 1.0) {
+    if (raft_refinement_) {
+      uint32_t k0        = static_cast<uint32_t>(refine_ratio_ * k);
+      // auto distances_tmp = raft::make_host_matrix<float, IdxT>(batch_size, k0);
+      // auto candidates    = raft::make_host_matrix<IdxT, IdxT>(batch_size, k0);
+      auto distances_tmp = raft::make_device_matrix<float, IdxT>(gpu_resource_->getRaftHandle(device_), batch_size, k0);
+      auto candidates    = raft::make_device_matrix<IdxT, IdxT>(gpu_resource_->getRaftHandle(device_), batch_size, k0);
+      index_->search(batch_size,
+                     queries,
+                     k0,
+                     distances_tmp.data_handle(),
+                     candidates.data_handle(),
+                     this->search_params_.get());
+      // auto queries_v = raft::make_host_matrix_view<const T, IdxT>(queries, batch_size, index_->d);
+
+
+      // auto dataset_v = raft::make_host_matrix_view<const T, faiss::idx_t>(
+      //   this->dataset_, index_->ntotal, index_->d);
+
+      // auto neighbors_v =
+      //   raft::make_host_matrix_view<IdxT, IdxT>(reinterpret_cast<IdxT*>(neighbors), batch_size, k);
+      // auto distances_v = raft::make_host_matrix_view<float, IdxT>(distances, batch_size, k);
+
+      // raft::runtime::neighbors::refine(gpu_resource_->getRaftHandle(device_),
+      //                                  dataset_v,
+      //                                  queries_v,
+      //                                  candidates.view(),
+      //                                  neighbors_v,
+      //                                  distances_v,
+      //                                  metric_faiss_to_raft(index_->metric_type));
+
+      auto queries_host    = raft::make_host_matrix<T, IdxT>(batch_size, index_->d);
+      auto candidates_host = raft::make_host_matrix<IdxT, IdxT>(batch_size, k0);
+      auto neighbors_host  = raft::make_host_matrix<IdxT, IdxT>(batch_size, k);
+      auto distances_host  = raft::make_host_matrix<float, IdxT>(batch_size, k);
+      auto dataset_v = raft::make_host_matrix_view<const T, faiss::idx_t>(
+        this->dataset_, index_->ntotal, index_->d);
+
+      auto handle_ = gpu_resource_->getRaftHandle(device_);
+
+      raft::copy(queries_host.data_handle(), queries, queries_host.size(), stream);
```

Comment:
I am allowing queries to be on both -- host and device. When the queries are on host and we have refinement ratios > 1, use FAISS' refinement methods. When they are on device and we have refinement enabled, we use raft's refinement API because FAISS' `IndexRefineFlat` requires the queries and distances and candidates to be on host. The latter gives us a more apples to apples comparison with RAFT because:
1. No unwanted copies of queries to device in FAISS' internal impl
2. Avoid unnecessary overhead (if any) due to FAISS' refinement methods